### PR TITLE
MNT: upgrade ruff to v0.5

### LINKED ISSUE #16621: MNT: migration to ruff 0.5.0
This blogpost just dropped: https://astral.sh/blog/ruff-v0.5.0#migrating-to-v05
Since there's a migration guide I expect the upgrade might not be smooth enough for a pre-commit.ci bot PR to handle correctly.
In particular these sections might need a careful look:
- https://astral.sh/blog/ruff-v0.5.0#rule-remappings
- https://astral.sh/blog/ruff-v0.5.0#stabilized-rules

However v0.5.0 is not available just yet so I can't confirm or deny there's an actual problem here.

I'll try it out tomorrow unless someone else beats me to it !

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #16621
I've separated changes into commits of decreasing safety; autofixes come first, then manual ones.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -154,9 +154,9 @@ def __init__(cls, name, bases, members, **kwds):
         pdict = {}
         for base in bases:
             for tbase in base.__mro__:
```
Comment (nstarman) [inline astropy/modeling/core.py:156]: Can this be compressed into another loop comprehension?

**Review by nstarman — APPROVED** [review]
Looks good for units! Thanks.

File: astropy/io/votable/tree.py [hunk 2]
```diff
@@ -4285,8 +4285,8 @@ def to_xml(
         }
         kwargs.update(self._get_version_checks())
 
-        with util.convert_to_writable_filelike(fd, compressed=compressed) as fd:
-            w = XMLWriter(fd)
+        with util.convert_to_writable_filelike(fd, compressed=compressed) as fh:
```
Comment (pllim) [inline astropy/io/votable/tree.py:4288]: I am really not sure but does `with something(fd, ...) as fd` allow the old `fd` to be garbage collected faster?

File: astropy/modeling/core.py [hunk 3]
```diff
@@ -668,11 +668,6 @@ class Model(metaclass=_ModelMeta):
     in the class body.
     """
 
-    n_inputs = 0
```
Comment (pllim) [inline astropy/modeling/core.py:671]: This would need inputs from @nden or @WilliamJamieson . Surely they were done as such for a reason, maybe to ensure it returns 0 in some cases?

File: astropy/wcs/utils.py [hunk 4]
```diff
@@ -1196,7 +1196,7 @@ def fit_wcs_from_points(
         wcs.wcs.cd = wcs.wcs.pc
         wcs.wcs.__delattr__("pc")
 
-    if (type(sip_degree) != type(None)) and (type(sip_degree) != int):
+    if (sip_degree is not None) and (type(sip_degree) != int):
```
Comment (pllim) [inline astropy/wcs/utils.py:1199]: Let's ping @mcara in case there is something subtle that would require comparing `type` directly like that.

File: astropy/io/votable/tree.py (hunk 2 above)
Comment (neutrinoceros) [reply astropy/io/votable/tree.py:4288]: I would imagine that `fh` should hold a strong reference on `fd`, so my guess is no.

File: astropy/modeling/core.py (hunk 3 above)
Comment (neutrinoceros) [reply astropy/modeling/core.py:671]: Actually this is directly breaking the properties with the same names, who rely on the class attribute as a default. Sorry, I should have read this more carefully. I think I can simply give the class attributes a different (private) name to satisfy the linter.

File: astropy/modeling/core.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/modeling/core.py:156]: looking at it more carefully I don't think it should; I think the more logical way to express what's going on here is actually:
```python
        pdict = {}
        for base in bases:
            for tbase in base.__mro__:
                if issubclass(tbase, Model):
                    pdict |= cls._parameters_
                    break
```
(unless `cls._parameters_` is somehow allowed to mutate in between iterations but that would seem extremely fragile)

File: astropy/modeling/core.py [hunk 5]
```diff
@@ -4492,7 +4458,7 @@ def _custom_model_inputs(func):
             special_params[param.name] = param.default
         elif param.name in settable:
             settable_params[param.name] = param.default
-        elif param.name in properties:
+        elif param.name in properties or param.name in reserved:
```
Comment (neutrinoceros) [inline astropy/modeling/core.py:4461]: this is to preserve exact backward compatibility as we have a test checking for this specific use case. Maybe the correct fix should be to add `"n_inputs"` to `special` and update the test instead. Maybe this change would make more sense in a dedicated PR.

Comment (neutrinoceros) [reply astropy/modeling/core.py:4461]: moved to #16633

**Comment by neutrinoceros** [comment]
making this a draft until I have a proper fix for #16633

**Comment by neutrinoceros** [comment]
This should not be independent from #16633, undrafting.

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -60,7 +60,7 @@ repos:
           - tomli
 
   - repo: https://github.com/astral-sh/ruff-pre-commit
-    rev: "v0.4.7"
+    rev: "v0.5.0"
     hooks:
       - id: ruff
         args: ["--fix", "--show-fixes"]
```
Comment (eerovaher) [inline .pre-commit-config.yaml:66]: [Have you checked if we still need `--show-fixes`?](https://astral.sh/blog/ruff-v0.5.0#output-format-now-defaults-to-full)

File: astropy/config/tests/test_configs.py [hunk 7]
```diff
@@ -353,7 +353,7 @@ class Conf(ConfigNamespace):
     with open(f, "wb") as fd:
         apycfg.write(fd)
     with open(f, encoding="utf-8") as fd:
-        lns = [x.strip() for x in fd.readlines()]
+        lns = [x.strip() for x in fd]
```
Comment (eerovaher) [inline astropy/config/tests/test_configs.py:356]: Is `strip()` only used for getting rid of the line endings? There's a few more analogous changes, but I'm only highlighting it here.

File: astropy/io/fits/column.py [hunk 8]
```diff
@@ -1047,14 +1047,16 @@ def _verify_keywords(
...
-        ]:
-            if v is not None and v != "":
-                valid[k] = v
+        valid |= {
+            k: v
+            for k, v in [
+                ("name", name),
+                ("unit", unit),
+                ("bscale", bscale),
+                ("bzero", bzero),
```
Comment (eerovaher) [inline astropy/io/fits/column.py:1056]: Have you checked if the `list` fits in a single line if the trailing comma is removed?

File: astropy/coordinates/baseframe.py [hunk 9]
```diff
@@ -273,9 +273,9 @@ def __init_subclass__(cls, **kwargs):
             if issubclass(basecls, BaseCoordinateFrame):
                 frame_attrs.update(basecls.frame_attributes)
 
-        for k, v in cls.__dict__.items():
-            if isinstance(v, Attribute):
-                frame_attrs[k] = v
+        frame_attrs |= {
+            k: v for k, v in cls.__dict__.items() if isinstance(v, Attribute)
```
Comment (eerovaher) [inline astropy/coordinates/baseframe.py:277]:
```suggestion
            k: v for k, v in vars(cls).items() if isinstance(v, Attribute)
```

File: astropy/io/fits/util.py [hunk 10]
```diff
@@ -302,12 +302,9 @@ def isreadable(f):
     if not hasattr(f, "read"):
         return False
 
-    if hasattr(f, "mode") and not any(c in f.mode for c in "r+"):
-        return False
-
     # Not closed, has a 'read()' method, and either has no known mode or a
     # readable mode--should be good enough to assume 'readable'
-    return True
+    return not (hasattr(f, "mode") and not any(c in f.mode for c in "r+"))
```
Comment (eerovaher) [inline astropy/io/fits/util.py:307]:
```suggestion
    return (not hasattr(f, "mode")) or any(c in f.mode for c in "r+")
```

File: astropy/io/fits/util.py [hunk 11]
```diff
@@ -325,12 +322,9 @@ def iswritable(f):
     if not hasattr(f, "write"):
         return False
 
-    if hasattr(f, "mode") and not any(c in f.mode for c in "wa+"):
-        return False
-
     # Note closed, has a 'write()' method, and either has no known mode or a
     # mode that supports writing--should be good enough to assume 'writable'
-    return True
+    return not (hasattr(f, "mode") and not any(c in f.mode for c in "wa+"))
```
Comment (eerovaher) [inline astropy/io/fits/util.py:327]:
```suggestion
    return (not hasattr(f, "mode")) or any(c in f.mode for c in "wa+")
```

File: astropy/io/votable/xmlutil.py [hunk 12]
```diff
@@ -60,9 +60,7 @@ def check_token(token, attr_name, config=None, pos=None):
 
     As defined by XML Schema Part 2.
     """
-    if token is not None and not xml_check.check_token(token):
-        return False
-    return True
+    return not (token is not None and not xml_check.check_token(token))
```
Comment (eerovaher) [inline astropy/io/votable/xmlutil.py:63]:
```suggestion
    return token is None or xml_check.check_token(token)
```

File: astropy/modeling/fitting.py [hunk 13]
```diff
@@ -411,9 +411,7 @@ def _is_invertible(m):
         """Check if inverse of matrix can be obtained."""
         if m.shape[0] != m.shape[1]:
             return False
-        if np.linalg.matrix_rank(m) < m.shape[0]:
-            return False
-        return True
+        return not np.linalg.matrix_rank(m) < m.shape[0]
```
Comment (eerovaher) [inline astropy/modeling/fitting.py:414]:
```suggestion
        return np.linalg.matrix_rank(m) >= m.shape[0]
```
If there is some `nan` weirdness going on then it should be mentioned in a comment or in the docstring.

File: astropy/table/tests/test_groups.py [hunk 14]
```diff
@@ -636,9 +636,7 @@ def test_column_filter():
     """
 
     def all_positive(column):
-        if np.any(column < 0):
-            return False
-        return True
+        return not np.any(column < 0)
```
Comment (eerovaher) [inline astropy/table/tests/test_groups.py:639]: A rewrite of the code should be
```suggestion
        return np.all(column >= 0)
```
as long as we don't have to worry about `nan` (it should be mentioned in a comment if that is a concern). But looking at the function name the correct code would actually be
```suggestion
        return np.all(column > 0)  # no NaN to worry about
```
or
```suggestion
        return not np.any(column <= 0)  # can handle NaN
```

File: astropy/units/format/cds.py [hunk 15]
```diff
@@ -68,13 +68,7 @@ def _units(cls) -> dict[str, UnitBase]:
...
         from astropy.units import cds
 
-        names = {}
-
-        for key, val in cds.__dict__.items():
-            if isinstance(val, u.UnitBase):
-                names[key] = val
-
-        return names
+        return {k: v for k, v in cds.__dict__.items() if isinstance(v, u.UnitBase)}
```
Comment (eerovaher) [inline astropy/units/format/cds.py:71]:
```suggestion
        return {k: v for k, v in vars(cds).items() if isinstance(v, u.UnitBase)}
```

File: .pre-commit-config.yaml (hunk 6 above)
Comment (neutrinoceros) [reply .pre-commit-config.yaml:66]: We do. Here's a screenshot of ruff autofixing the same lint twice, first with then without this flag
Screenshot 2024-06-28 at 21 57 06

File: astropy/config/tests/test_configs.py (hunk 7 above)
Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py:356]: I don't know, but I wouldn't want this PR to mutate into an unwarranted refactor.

File: astropy/io/fits/column.py (hunk 8 above)
Comment (neutrinoceros) [reply astropy/io/fits/column.py:1056]: I did. It doesn't.

File: astropy/table/tests/test_groups.py (hunk 14 above)
Comment (neutrinoceros) [reply astropy/table/tests/test_groups.py:639]: I'll go with your first suggestion as the other ones are not strictly equivalent.
Some definitions of "positive" include 0 so I'm not willing to bet the current behavior is buggy.

File: astropy/config/tests/test_configs.py (hunk 7 above)
Comment (eerovaher) [reply astropy/config/tests/test_configs.py:356]: On the contrary, now is a very good time for some unwarranted refactoring. Ruff is highlighting code for us that we should look at and might want to improve. If all we ever do is make minimal changes so that Ruff stops complaining then we might still have code that should be improved, but we can't use Ruff to find it anymore.

If addressing the relevant Ruff rule seems to be too complicated or too time consuming then it would be better to add a temporary ignore to `.ruff.toml` and to properly address the rule at some later point.

File: astropy/modeling/fitting.py [hunk 16]
```diff
@@ -411,9 +411,7 @@ def _is_invertible(m):
         """Check if inverse of matrix can be obtained."""
         if m.shape[0] != m.shape[1]:
             return False
-        if np.linalg.matrix_rank(m) < m.shape[0]:
-            return False
-        return True
+        return np.linalg.matrix_rank(m) >= m.shape[0]
```
Comment (eerovaher) [inline astropy/modeling/fitting.py:414]: This could be refactored even more aggressively:
```suggestion
        return m.shape[0] == m.shape[1] <= np.linalg.matrix_rank(m)
```
or maybe (more verbosely)
```suggestion
        return m.shape[0] == m.shape[1] and np.linalg.matrix_rank(m) >= m.shape[0]
```

File: astropy/config/tests/test_configs.py (hunk 7 above)
Comment (dhomeier) [reply astropy/config/tests/test_configs.py:356]: What is your suggestion for substituting `strip()` then? Or does the `in fd` already iterate without the endings?

File: astropy/io/ascii/qdp.py [hunk 17]
```diff
@@ -138,7 +138,7 @@ def _get_lines_from_file(qdp_file):
         lines = qdp_file.split("\n")
     elif isinstance(qdp_file, str):
         with open(qdp_file) as fobj:
-            lines = [line.strip() for line in fobj.readlines()]
+            lines = [line.strip() for line in fobj]
```
Comment (dhomeier) [inline astropy/io/ascii/qdp.py:141]: In this case it passes tests with `line.rstrip("\n")`, so (leading) blanks are not an issue.

File: astropy/config/tests/test_configs.py (hunk 7 above)
Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py:356]: I imagine what @eerovaher has in mind would be something like:
```python
lns = fd.read().splitlines()
```
which does the same thing but is clearer. I'll give it a spin.

File: astropy/io/fits/scripts/fitsdiff.py [hunk 18]
```diff
@@ -80,7 +80,8 @@ def __call__(self, parser, namespace, values, option_string=None):
                 log.warning(f"{self.dest} argument {value} does not exist")
                 return
             try:
-                values = [v.strip() for v in open(value).readlines()]
+                with open(value) as fd:
+                    values = fd.read().splitlines()
```
Comment (neutrinoceros) [inline astropy/io/fits/scripts/fitsdiff.py:84]: I should note that I refactored this manually for consistency with other similarly affected snippets, but this particular one isn't covered by tests, and the new proposed code isn't strictly equivalent in that it reads the whole file eagerly. Happy to revert if this feels risky to package maintainers.

File: astropy/nddata/mixins/tests/test_ndarithmetic.py [hunk 19]
```diff
@@ -1181,9 +1181,7 @@ def meta_fun_func(meta1, meta2, take="first"):
 
 def test_arithmetics_wcs_func():
     def wcs_comp_func(wcs1, wcs2, tolerance=0.1):
-        if tolerance < 0.01:
-            return False
-        return True
+        return tolerance >= 0.01
```
Comment (taldcroft) [inline astropy/nddata/mixins/tests/test_ndarithmetic.py:1184]: Like the other comment, this is potentially changing the output type here if `tolerance` is a numpy array scalar. Maybe it doesn't matter but ruff changes should avoid subtle code changes, no?

File: astropy/table/tests/test_groups.py [hunk 20]
```diff
@@ -636,9 +636,7 @@ def test_column_filter():
     """
 
     def all_positive(column):
-        if np.any(column < 0):
-            return False
-        return True
+        return np.all(column >= 0)
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/table/tests/test_groups.py:639]: or ``return bool(np.any(column >= 0)``

[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/table/tests/test_groups.py:639]: Then a followup PR can change the type.

File: astropy/config/tests/test_configs.py [hunk 21]
```diff
@@ -353,7 +353,7 @@ class Conf(ConfigNamespace):
     with open(f, "wb") as fd:
         apycfg.write(fd)
     with open(f, encoding="utf-8") as fd:
-        lns = [x.strip() for x in fd.readlines()]
+        lns = fd.read().splitlines()
```
[the remark this answers is not in the document]
Comment (dhomeier) [reply astropy/config/tests/test_configs.py:356]: _Should_ not matter since it was likely just for removing the linebreaks (see previous discussion); I think other blanks will be stripped by the column parser anyway.

File: astropy/nddata/mixins/tests/test_ndarithmetic.py (hunk 19 above)
Comment (neutrinoceros) [reply astropy/nddata/mixins/tests/test_ndarithmetic.py:1184]: This one is semi manual anyway, ruff doesn't know how to simplify `not` VS inequalities. I can convert the output to a Python bool manually for better safety.
You're right that ruff should avoid this kind of changes manually but I think what it did here is right >99% of the time and it couldn't statically guess about the remaining epsilon% so in my honest opinion it's fine to keep this rule.

File: astropy/config/tests/test_configs.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py:356]: @taldcroft sorry we ended up mixing auto formatting with manual refactors. To be clear here's our process

```
[x.strip() for x in fd.readlines()]
⬇️ (ruff)
[x.strip() for x in fd]
⬇️ (manual refactor)
fd.read().splitlines()
```

**Comment by pllim** [comment]
The bot has started trying to auto update (#16645). Is there enough consensus to wrap this up soon? Thanks, all!

File: astropy/config/tests/test_configs.py (hunk 7 above)
Comment (eerovaher) [reply astropy/config/tests/test_configs.py]: What I actually had in mind was
```patch
-    with open(f, encoding="utf-8") as fd:
-        lns = [x.strip() for x in fd.readlines()]
-
-    assert "tstnmo = op2" in lns
+    assert "tstnmo = op2" in f.read_text().splitlines()
```
but `f` looks small enough that maybe we could check all of it.

File: astropy/io/ascii/qdp.py [hunk 22]
```diff
@@ -138,7 +138,7 @@ def _get_lines_from_file(qdp_file):
         lines = qdp_file.split("\n")
     elif isinstance(qdp_file, str):
         with open(qdp_file) as fobj:
-            lines = [line.strip() for line in fobj.readlines()]
+            lines = fobj.read().splitlines()
```
[the remark this answers is not in the document]
Comment (eerovaher) [reply astropy/io/ascii/qdp.py:141]: Are you saying that `[line.strip() for line in fobj.readlines()]` is not always equivalent to `fobj.read().splitlines()` or are you saying that if `fobj` is a QDP file then the new code is different? The former is true, but the latter might not be, depending on what the format is like and how the `astropy` code reading it is implemented (but I don't know about these details).

File: astropy/table/tests/test_groups.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (eerovaher) [reply astropy/table/tests/test_groups.py]: `all_positive()` is not needed, we could instead write
```patch
-    c2 = tg["c"].groups.filter(all_positive)
+    c2 = tg["c"].groups.filter(lambda column: np.all(column >= 0))
```
and that would also prevent us from discussing whether `all_positive()` should return `np.all(columns >= 0)` or `np.all(columns > 0)`.

**Comment by neutrinoceros** [comment]
If anyone wants to oppose merging the current state because one of the changes is suspected to subtly change behavior, I can also revert them and ignore the rules deem too problematic. This was meant as a quick PR to avoid noise from the bot. Refactors can wait as far as I'm concerned. Just let me know !

File: astropy/nddata/mixins/tests/test_ndarithmetic.py (hunk 19 above)
Comment (eerovaher) [reply astropy/nddata/mixins/tests/test_ndarithmetic.py]: This function is defined inside a test and is not even accessible from anywhere else. If the test still succeeds then the replacement was good enough. We don't need to worry about API stability here.

File: astropy/io/fits/scripts/fitsdiff.py (hunk 18 above)
Comment (neutrinoceros) [reply astropy/io/fits/scripts/fitsdiff.py:84]: Well, technically, strip a line before it's stored would still be more memory effective than storing everything, then strip. The first implementation barely requires more memory than the final result, while this refactored version might consume up to a bit over twice as much, but in practice it really shouldn't matter.

File: astropy/config/tests/test_configs.py (hunk 7 above)
Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py]: Assert statements shouldn't be mixed with business logic as they can (in principle) be "optimized" away with `python -O`

Comment (eerovaher) [reply astropy/config/tests/test_configs.py]: But this is a test.

Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py]: oh my, I get so lost in the maze of the discussion

Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py]: done !

File: astropy/io/fits/scripts/fitsdiff.py (hunk 18 above)
Comment (eerovaher) [reply astropy/io/fits/scripts/fitsdiff.py:84]: `readlines()` returns a `list`, so that's reading eagerly too.

File: astropy/io/ascii/qdp.py (hunk 17 above)
Comment (taldcroft) [reply astropy/io/ascii/qdp.py]: Without getting into a detailed review of the code and format specs, we cannot assumes that the test inputs fully cover all possibilities. In other words, in a PR like this we really should stick with changes that do not obviously impact the code behavior.

File: astropy/io/ascii/qdp.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (taldcroft) [reply astropy/io/ascii/qdp.py:141]: If the file has leading or trailing spaces then the two are not equivalent. I do not have time to think about QDP and the code in order to make ruff happy.

[the remark this answers is not in the document]
Comment (dhomeier) [reply astropy/io/ascii/qdp.py:141]: As per my comment above, the column parser would likely still take care of remaining blanks, but that assumption could subtly fail if you have different column separators like `,` or `\t`. Dunno enough about the QDP format to say with certainty here.

File: astropy/io/fits/scripts/fitsdiff.py (hunk 18 above)
Comment (dhomeier) [reply astropy/io/fits/scripts/fitsdiff.py:84]: This is not even reading a data file, just a list of header keys or comments to be ignored, so should not be any problem to read the whole thing at once.

Comment (dhomeier) [reply astropy/io/fits/scripts/fitsdiff.py:84]: But comparing with how the values from the command line are set in L92, the previous code would be robust against files with blanks (e.g. for alignment of the entries), while this version is not.

**Comment by neutrinoceros** [comment]
Conflicts have started showing up and the conversation is getting way beyond the originally intended scope - that's on me, I should have avoided doing all subpackages at once.
I just reverted all controversial refactors so this can hopefully be merged quickly. I will reopen new and smaller PRs with just those refactors and encourage everyone to resolve their threads here.

File: astropy/config/tests/test_configs.py (hunk 7 above)
Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py:356]: This refactor was moved to #16656. Please consider closing this thread.

File: astropy/io/ascii/qdp.py (hunk 17 above)
Comment (neutrinoceros) [reply astropy/io/ascii/qdp.py:141]: This refactor was moved to #16656. Please consider closing this thread.

File: astropy/io/fits/scripts/fitsdiff.py (hunk 18 above)
Comment (neutrinoceros) [reply astropy/io/fits/scripts/fitsdiff.py]: This refactor was moved to #16656. Please consider closing this thread.

File: astropy/config/tests/test_configs.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/config/tests/test_configs.py]: This refactor was moved to #16656. Please consider closing this thread.

File: astropy/io/ascii/qdp.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/io/ascii/qdp.py]: This refactor was moved to #16656. Please consider closing this thread.

File: astropy/table/tests/test_groups.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/table/tests/test_groups.py]: This refactor was moved to #16655. Please consider closing this thread.

File: astropy/nddata/mixins/tests/test_ndarithmetic.py (hunk 19 above)
Comment (neutrinoceros) [reply astropy/nddata/mixins/tests/test_ndarithmetic.py]: This refactor was reverted entirely. Please consider closing this thread.

File: astropy/utils/data_info.py [hunk 23]
```diff
@@ -490,17 +490,17 @@ class = Column
             info["name"] = name
 
         options = option if isinstance(option, (list, tuple)) else [option]
-        for option in options:
-            if isinstance(option, str):
-                if hasattr(self, "info_summary_" + option):
-                    option = getattr(self, "info_summary_" + option)
+        for opt in options:
```
Comment (mhvk) [inline astropy/utils/data_info.py:493]: This is a place where ruff just makes code less clear: in the line above, `option` has effectively been made irrelevant, and it is nice to have `for singular in plural` -- and now the someone reading the code has to deal with an abbreviation (less clear) and more verbose exception.

If we go this route, maybe at least get rid of the abbreviation by appending an underscore, and/or do
```
option = option if isinstance(option, (list, tuple)) else [option]
for option_ in option:
    ...
```

p.s. Not a big enough deal to stop this, but it explains my fatigue with the stylistic changes...

**Review by mhvk — APPROVED** [review]
units and utils are fine (though see comment on units)

File: astropy/utils/data_info.py (hunk 23 above)
Comment (neutrinoceros) [reply astropy/utils/data_info.py:493]: For the record, this rule (PLR1704) is from pylint (which I personally feel is generally noisy). Though the refactor is manual and I don't want to hurt readability. I'll use `option_` as suggested.

File: astropy/modeling/core.py (hunk 3 above)
Comment (pllim) [reply astropy/modeling/core.py:671]: @nden , are you okay with this?

**Comment by neutrinoceros** [comment]
seeing that rule NPY201 was once again updated in ruff 0.5.1 (but with no effect on astropy), I bumped again so it's clearer in the future that this release doesn't cause new lint errors.

**Comment by neutrinoceros** [comment]
and rebased for docs builds

**Review by nstarman — APPROVED** [review]
:shipit:

File: astropy/modeling/core.py (hunk 3 above)
Comment (pllim) [reply astropy/modeling/core.py:671]: @WilliamJamieson ?

**Comment by pllim** [comment]
tl;dr -- Are the discussions with @eerovaher and @taldcroft resolved?

File: astropy/modeling/core.py (hunk 3 above)
Comment (neutrinoceros) [reply astropy/modeling/core.py:671]: Maybe I should point out that those attributes were already not accessible at runtime as they are shadowed by properties with the same names (which is what ruff is pointing out).

Comment (neutrinoceros) [reply astropy/modeling/core.py:671]: I should also link #16633 were I did a more thorough cleanup

Comment (pllim) [reply astropy/modeling/core.py:671]: Given https://github.com/astropy/astropy/pull/16633 , should we just not touch modeling here?

Comment (neutrinoceros) [reply astropy/modeling/core.py:671]: Well I have to do something about it here to pass linting, and this bit is the same so there should not be any conflict.

**Comment by eerovaher** [comment]
I think the discussions have concluded, but it would be better to merge #16633 first to reduce clutter in the diff here, and to re-implement here the changes that were moved to other pull requests (and essentially approved there).

**Comment by taldcroft** [comment]
I see all the `strip()` method calls seem to be back in place, so I don't have any objections to this. I have not re-reviewed the whole thing in detail so I'm not formally "Approving".

**Comment by neutrinoceros** [comment]
bumped to ruff 0.5.2 (noop)

**Comment by neutrinoceros** [comment]
bumped to ruff 0.5.3 (noop)

File: astropy/wcs/utils.py (hunk 4 above)
Comment (nstarman) [reply astropy/wcs/utils.py:1199]: @mcara, does this look good?

**Comment by neutrinoceros** [comment]
bumped to ruff 0.5.4 (noop)

**Comment by neutrinoceros** [comment]
bumped to ruff 0.5.5 (noop)

**Comment by nstarman** [comment]
Who else should approve this PR before we merge?

**Comment by pllim** [comment]
I thought this is blocked by #16633 as per https://github.com/astropy/astropy/pull/16630#issuecomment-2211330561 ?

**Comment by neutrinoceros** [comment]
Not anymore : https://github.com/astropy/astropy/pull/16633#issuecomment-2228513077

File: astropy/coordinates/baseframe.py (hunk 9 above)
Comment (eerovaher) [reply astropy/coordinates/baseframe.py:277]: Using `vars` here would allow this to fit on one line.

**Review by eerovaher — COMMENTED** [review]
I would still prefer #16633 to be dealt with before merging this pull request, but we should merge this before the end of next week, so if #16633 is not reviewed quickly enough then we should move on with this regardless.

Have the changes that were moved to separate pull requests and approved there been moved back here? There's at least one change in `coordinates` that hasn't.

File: astropy/coordinates/baseframe.py (hunk 9 above)
Comment (neutrinoceros) [reply astropy/coordinates/baseframe.py:277]: Quoting @mhvk from https://github.com/astropy/astropy/pull/16652

> Overall, my conclusion is to let this be -- quite a bit of churn for no tangible benefit.

I'd like to avoid going through another round of discussion if you don't mind too much, this *stylistic* PR has been stalling for a while already.

**Comment by neutrinoceros** [comment]
> Have the changes that were moved to separate pull requests and approved there been moved back here? 

I haven't moved anything back. Here's a best-effort attempt to find these PRs back:
- #16652 (rejected)
- #16653 (approved by @mhvk)
- #16656 (same as above)
- #16655 (already merged)

> There's at least one change in coordinates that hasn't.

Can you be more specific ? I couldn't find that one.

File: astropy/coordinates/baseframe.py (hunk 9 above)
Comment (eerovaher) [reply astropy/coordinates/baseframe.py:277]: We are editing this anyways, and editing it differently does not cause additional code churn.

**Review by mwcraig — APPROVED** [review]

**Comment by neutrinoceros** [comment]
bumped to ruff 0.5.6 (noop)

**Comment by nstarman** [comment]
Ping @eerovaher :)

**Comment by eerovaher** [comment]
All the pull requests that were split off from this one should be closed and the changes implemented here, with the exception of #16633 because that is the only one where it is not obvious how to proceed. From what I remember this pull request was only waiting for the decision regarding #16633 when the other pull requests were opened, so opening them just made the Ruff upgrade more difficult to review.

**Comment by dhomeier** [comment]
> * [TST: clarify a test where str.strip is called while iterating over file content #16656](https://github.com/astropy/astropy/pull/16656) (same as above)

Which "above" are you referring to here? As I understand https://github.com/astropy/astropy/pull/16656#issuecomment-2203094101, only 2197067 should be kept from that PR, but so far the other changes have not been reverted there.

**Comment by neutrinoceros** [comment]
@dhomeier by "same as above" I meant "approved by Marten". You're right that only the last commit from this PR is to be kept, all other commits are originated from the present PR (though maybe out of sync at the moment).

@eerovaher I still don't know what coordinates change you meant to refer to in https://github.com/astropy/astropy/pull/16630#pullrequestreview-2200052176, though I'll incorporate back the three PRs I listed in https://github.com/astropy/astropy/pull/16630#issuecomment-2251178395 now

File: astropy/modeling/core.py [hunk 24]
```diff
@@ -808,7 +808,8 @@ def outputs(self, val):
         self._outputs = val
 
     @property
-    def n_inputs(self):
+    def n_inputs(self):  # noqa: F811 # see https://github.com/astropy/astropy/pull/16633
```
Comment (eerovaher) [inline astropy/modeling/core.py:811]: All the temporary Ruff configuration belongs in `.ruff.toml`. The configuration in `pyproject.toml` and all uses of `noqa` are meant to be permanent.

File: astropy/coordinates/baseframe.py (hunk 9 above)
Comment (eerovaher) [reply astropy/coordinates/baseframe.py:277]: This code could be made much more compact by using `vars()`.

File: astropy/modeling/core.py (hunk 24 above)
Comment (neutrinoceros) [reply astropy/modeling/core.py:811]: I really cannot see both not make #16633 blocking and not have a noqa here. Any suggestion ?

File: astropy/coordinates/baseframe.py (hunk 9 above)
Comment (neutrinoceros) [reply astropy/coordinates/baseframe.py:277]: looks like you *do* mind... I'll push this now if that's what it takes.

File: astropy/modeling/core.py (hunk 24 above)
Comment (eerovaher) [reply astropy/modeling/core.py:811]: Add the rule to the `modeling` section in `.ruff.toml`. Whether we should update the code to enforce the rule or ignore the rule with noqas should be the outcome of #16633.

Comment (neutrinoceros) [reply astropy/modeling/core.py]: done ! I rebased the branch to avoid back-and-forth in the history

**Comment by neutrinoceros** [comment]
Ruff 0.6 is just out and also comes with a migration guide. I'd like to avoid keeping this PR around forever so I will not be bumping further than 0.5.7 here.

**Comment by neutrinoceros** [comment]
(still, adding some notes for the *next* PR since I'm reading the migration guide anyway)

**Comment by pllim** [comment]
Should we just skip 0.5 and go straight to 0.6? 👀 😬

**Comment by nstarman** [comment]
It looks like all the blockers for this PR are resolved / have their own PR.
IMO merge this in then tackle v0.6.

**Review by mhvk — APPROVED** [review]
I looked over everything again and think we should just get this in; there is nothing remotely controversial left.

**Comment by nstarman** [comment]
Thanks @neutrinoceros for championing this PR!
